rosbag2_cpp/src/rosbag2_cpp/clocks/time_controller_clock.cpp

```diff
@@ -112,9 +121,10 @@ rcutils_time_point_value_t TimeControllerClock::now() const
 
 bool TimeControllerClock::sleep_until(rcutils_time_point_value_t until)
 {
+  const auto until_timeout = impl_->now_fn() + impl_->sleep_timeout;
   {
     TSAUniqueLock lock(impl_->state_mutex);
-    const auto steady_until = impl_->ros_to_steady(until);
+    const auto steady_until = std::min(impl_->ros_to_steady(until), until_timeout);
```

Comment:
Yes - I was planning to `if (pause) { wait_for(sleep_timeout) }`, you'll notice that in this PR `pause` doesn't exist yet. I wanted to also support the clock publisher use case https://github.com/ros2/rosbag2/pull/695/files where I need  to call `spin()` frequently enough for the node timers to run.

I'm now thinking maybe I should just create a separate thread for that spin, it's probably a simpler approach. In which case, I can just do a single PR for `pause/resume` instead of breaking this up into two parts.

Reply:
@emersonknapp I think moving publisher for `/clock` topic in separate thread would be a better idea.